File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_map<string , int> mp;
        int n = wordList.size();
        for(int i = 0 ; i < n ; i++){
            mp[wordList[i]] = i;
        }
        mp.erase(beginWord);
        queue<pair<string , int>> q;
        q.push({beginWord , 1});
        while(!q.empty()){
            auto [s , cnt] = q.front();
            q.pop();
            if(s == endWord) return cnt;
            int m = s.size();
            for(int j = 0 ; j < m ; j++){
                char ori = s[j];
                for(char i = 'a' ; i <= 'z' ; i++){
                    s[j] = i;
                    if(mp.find(s) != mp.end()){
                        q.push({s , cnt + 1});
                        mp.erase(s);
                    }
                }
                s[j] = ori; 
            }
        }
        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        int n = wordList.size();
        HashMap<String , Integer> mp = new HashMap<>();
        for(String s : wordList){
            mp.put(s , mp.getOrDefault(s , 0) + 1);
        }
        Queue<Pair<String , Integer>> q = new LinkedList<>();
        q.offer(new Pair<>(beginWord , 1));
        mp.remove(beginWord);
        while(!q.isEmpty()){
            String s = q.peek().getKey();
            int t = q.peek().getValue();
            q.poll();
            if(s.equals(endWord)) return t;
            int m = s.length();
            for(int i = 0 ; i < m ; i++){
                for(char j = 'a' ; j <= 'z' ; j++){
                    String ns = s.substring(0 , i) + j + s.substring(i + 1);
                    if(mp.containsKey(ns)){
                        q.offer(new Pair<>(ns , t + 1));
                        mp.remove(ns);
                    }
                }
            }
        }
        return 0;
    }
}